data/tutorials/language/0lg_09_env.md

```diff
@@ -0,0 +1,610 @@
+
+### Passing Tuples
+
+Consider this function:
+```ocaml
+# let space_cat (s, t) = s ^ " " ^ t;;
+val space_cat : string * string -> string = <fun>
+```
+
+It takes a pair of strings and returns a string.
+```ocaml
+# space_cat ("hello", "world");;
+- : string = "hello world"
+```
+
+Asking whether `space_cat` is a binary function or not makes sense. Morally, it takes two parameters: the two strings that sandwich the space character. Wrapped as a pair, they are a single piece of data, so `space_cat` has a single parameter. But it is that pair's elements that are relevant, not the pair as a whole. So morally, `space_cat` is a binary function. Syntactically, the call `space_cat ("hello", "world")` looks pretty much like a high school maths function call or a spreadsheet application, both having two parameters: `"hello"` and `"world"`.
+
+### Currying and Uncurrying
+
+In the two previous sections, two kinds of “multiple parameter” functions have been presented.
+- Functions returning a function, such as `( + )`
+- Functions taking a pair or a tuple as a parameter, such as `space_cat`
+
+Interestingly, both kinds of functions provide a way to pass several parameters while being functions with a single parameter. From this perspective, it makes sense to say: “All functions have a single argument.”
+
+This goes even further. Any function of the same kind can be translated into an equivalent function of the second kind, and conversely. Using OCaml support for higher-order functions, it is possible to define those transformations as functions.
+
+Here is the translation from the first kind to the second kind:
+```ocaml
+# fun f -> fun (x, y) -> f x y;;
+- : ('a -> 'b -> 'c) -> 'a * 'b -> 'c = <fun>
+```
+
+Here is the reverse translation:
+```ocaml
+# fun f -> fun x y -> f (x, y);;
+- : ('a * 'b -> 'c) -> 'a -> 'b -> 'c = <fun>
+```
+
+These translations are attributed to the 20th-century logician [Haskell Curry](https://en.wikipedia.org/wiki/Haskell_Curry). The second translation is called *currying* and the first is called *uncurrying*.
+
+From a typing perspective, this means that for any types `a`, `b`, and `c`, the following types are equivalent:
+- `a -> (b -> c)` &mdash; curried function type
+- `a * b -> c` &mdash; uncurried function type
+
+We will not dive any deeper into the details here, but this equivalence can be formally defined using _ad-hoc_ mathematics.
+
+It is rarely necessary to use the functions above. However, it is important to understand that changing a function may not need any code refactoring. A function can be given several equivalent forms and changed into another, either using refactoring or using a higher-order function. Since functions are values, currying and uncurrying are operations on those values.
```

Comment:
If this is rarely necessary, does it need to be in this relatively-beginner tutorial? Will it muddy the waters or is it important to know at this point of learning even if it's not used very much?

Reply:
Perhaps Currying & Uncurrying can be part of a more advanced tutorial on functions. If this is the first or second tutorial after the Getting Started ones, it might be too much detail for a relative beginner to OCaml and derail the happy path.

Reply:
Move 512-534 into another (TBD) document